### PR TITLE
feat(gitFileSystem): safer api

### PR DESCRIPTION
**Pull request by kishore03109**
## Problem

1. the presence of default arguments lead to sneaky bugs introduced due to caller unknowingly calling the wrong branch. 
2. there is a scary instance whereby the lack of update to staging-lite could occur, but the update to staging does not, leading to issues downstream. 

Closes [insert issue #]

## Solution
1. to keep things simple, we make all the args compulsory.
2. make staging update first before updating staging lite.

### CODE REVIEW

**Comment by kishore03109** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1042** <a href="https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1042?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1043** <a href="https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1043?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1044** <a href="https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1044?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #1045** <a href="https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1045?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #1046** <a href="https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1046?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1046?utm_source=stack-comment).

File: src/services/db/GitFileCommitService.ts
Comment (seaerchin) [inline src/services/db/GitFileCommitService.ts:95]: this can fail silently - we should probably just chain off this. alternatively, what we can do is just chain off the results sequentially and throw an error at the end (the behaviour now), which makes things clearer to read. 

also eliminates the issue of `staging-lite` updating but not `staging` cos if the initial call to `staging` fails, `staging-lite` won't even run

File: src/services/db/GitFileCommitService.ts
Comment (seaerchin) [inline src/services/db/GitFileCommitService.ts:152]: same issue as above; there's also a repeated pattern - we could probably just separate out the steps as functions and chain them together.

File: src/services/db/GitFileCommitService.ts
Comment (seaerchin) [inline src/services/db/GitFileCommitService.ts:1]: this file has alot of the same pattern of do some change -> check the result of that change -> update staging lite -> update github -> throw error if any surface.

we could probably separate those out and make this whole file clearer. that being said, we don't have to do it _now_. 

however, there is an issue where updates are allowed to silently fail at the push step. because our call returns a `result`, it won't throw by design and we need to check if it's recoverable or we should just fail fast (throw error).

File: src/services/db/GitFileSystemService.ts
Comment (seaerchin) [inline src/services/db/GitFileSystemService.ts:253]: nit: could we remove this `console.log`?

**Review by seaerchin — COMMENTED** [review]
the changes are fine, albeit repetitive. similar work should probably be abstracted out as a function to avoid slight behavioural drift over time that creates bugs. this is a maintainability issue but it's not too bad for now.

otoh, we have a possibility of a silent failure, where we commit to EFS but fail the push to github. this could potentially be very confusing and we could fail fast here or attempt to handle the error properly. no pref, as long as the error itself is handled.

File: src/services/db/GitFileCommitService.ts
Comment (kishore03109) [reply src/services/db/GitFileCommitService.ts:95]: Hey this was an intentional trade off we made as a team not to await for calls for they are expensive. But I do air your concerns regarding silent failures, this is the cause of the divergence that we are seeing. Dont intend to tackle it here!

For info only + outside of scope of pr: I tried adding an alarm for this, but it was TOO noisy, there is prob a way to make this such that we only throw alarms for this silent pushes, but havnt had time to invetigate

**Comment by kishore03109** [comment]
@seaerchin

> the changes are fine, albeit repetitive. similar work should probably be abstracted out as a function to avoid slight behavioural drift over time that creates bugs. this is a maintainability issue but it's not too bad for now.
 
I am quite curious how you would have structured this instead (but not today kek release first), maybe i book time w u after i come back to learn from you

>otoh, we have a possibility of a silent failure, where we commit to EFS but fail the push to github. this could potentially be very confusing and we could fail fast here or attempt to handle the error properly. no pref, as long as the error itself is handled.

I do think this is outside of the scope of this pr, since this is parity to what currently exists in production! thoughts?

File: src/services/db/GitFileCommitService.ts
Comment (kishore03109) [reply src/services/db/GitFileCommitService.ts]: can

File: src/services/db/GitFileCommitService.ts
Comment (seaerchin) [reply src/services/db/GitFileCommitService.ts:95]: we have a few different ways to tackle this tbh, just spitballing ideas - not an ask to implement. 

1. separate out the push to github and let the router handle this. the router can always send response early then fail later (no cost to user) if the push to github fails. 
2. because `neverthrow` is safe, we could just write a wrapper for retrying `Errors<T>`
3. just send an email on failure 

we probably want to decide what categories of errors we care about - transient errors vs persistent errors; in terms of our usage, transient is probably outages <10 mins in length (we don't have that much pushes tbvh) and persistent >=10. 

if transient, retry w/ exponential backoff will work well; if persistent, we probably want a notif + manual fix. to be honest, the divergence should arise only if we edit on github directly whilst there's a push failure so should be minimal but the problem is we got high barrier of entry to edit on efs (hence why we edit on github directly).

**Comment by seaerchin** [comment]
> I am quite curious how you would have structured this instead (but not today kek release first), maybe i book time w u after i come back to learn from you

ye sure

> I do think this is outside of the scope of this pr, since this is parity to what currently exists in production! thoughts?
if there's no easy win here then sure, we can just go with this

File: src/services/db/GitFileSystemService.ts
Comment (seaerchin) [reply src/services/db/GitFileSystemService.ts:253]: @kishore03109 pls remove thx

**Review by seaerchin — APPROVED** [review]
lgtm if no quick way of feeding back failures to us. 

do remove extra console log before merge.

**Comment by kishore03109** [comment]
### Merge activity

* **Dec 6, 5:51 AM**: @kishore03109 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/isomerpages/isomercms-backend/1046).

**Pull request merged**
